# Regresyon Modelleri için Hata Değerlendirme

Bu ödevde, verilen **bias** ve **weight** değerleriyle basit bir doğrusal regresyon modeli kurulacak; çalışanların deneyim yılına göre maaş tahmini yapılacak ve model başarısı **MSE**, **RMSE**, **MAE** metrikleriyle ölçülecektir.

## Veri Seti

Çalışanların deneyim yılı ($x$) ve maaş ($y$) bilgileri:

In [1]:
# Sayısal işlemler için NumPy kütüphanesini içe aktar
import numpy as np

# Tablo (DataFrame) işlemleri için Pandas kütüphanesini içe aktar
import pandas as pd

# Hazır hata metrikleri için sklearn fonksiyonlarını içe aktar
from sklearn.metrics import mean_squared_error, mean_absolute_error

# Deneyim yılı (x) ve maaş (y) verilerini DataFrame olarak oluştur
df = pd.DataFrame({
    "experience_year": [5, 7, 3, 3, 2, 7, 3, 10, 6, 4, 8, 1, 1, 9, 1],  # Bağımsız değişken: deneyim yılı
    "salary": [600, 900, 550, 500, 400, 950, 540, 1200, 900, 550, 1100, 460, 400, 1000, 380]  # Bağımlı değişken: maaş
})

# Oluşturulan veri setini ekrana yazdır
df

,experience_year,salary
0,5,600
1,7,900
2,3,550
3,3,500
4,2,400
5,7,950
6,3,540
7,10,1200
8,6,900
9,4,550


## Görev 1  Doğrusal Regresyon Model Denklemini Oluşturma

Verilen parametreler:
- **Bias (b)** = 275
- **Weight (w)** = 90

Basit doğrusal regresyon denklemi:

$$
\hat{y} = b + w x
$$

Bu değerlere göre model denklemi:

$$
\hat{y} = 275 + 90x
$$

**Yorum:** Deneyim yılı 0 iken tahmini maaş 275 birimdir. Her 1 yıllık deneyim artışı, tahmini maaşı 90 birim artırır.

In [2]:
# Modelin sabit terimi (bias / intercept)
bias = 275

# Deneyim yılının katsayısı (weight / slope)
weight = 90

# Oluşturulan doğrusal regresyon denklemini yazdır
print(f"Model equation: y' = {bias} + {weight}x")

Model equation: y' = 275 + 90x


## Görev 2 — Tüm Deneyim Yılları için Maaş Tahmini

Her satır için $\hat{y}_i = 275 + 90 \cdot x_i$ hesaplanır.

In [3]:
# Her satır için maaş tahmini hesapla: y' = bias + weight * x
df["salary_pred"] = bias + weight * df["experience_year"]

# Tahmin sütunu eklenmiş tabloyu göster
df

,experience_year,salary,salary_pred
0,5,600,725
1,7,900,905
2,3,550,545
3,3,500,545
4,2,400,455
5,7,950,905
6,3,540,545
7,10,1200,1175
8,6,900,815
9,4,550,635


## Görev 3 — MSE, RMSE, MAE Hesaplama

Önce hata, hata karesi ve mutlak hata sütunları oluşturulur; ardından metrikler hesaplanır.

$$
MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2
$$

$$
RMSE = \sqrt{MSE}
$$

$$
MAE = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|
$$

In [4]:
# Gerçek maaş ile tahmin arasındaki farkı (hata) hesapla
df["error"] = df["salary"] - df["salary_pred"]

# Hataların karesini al (MSE hesabı için)
df["squared_error"] = df["error"] ** 2

# Hataların mutlak değerini al (MAE hesabı için)
df["absolute_error"] = df["error"].abs()

# Hata sütunları eklenmiş tabloyu göster
df

,experience_year,salary,salary_pred,error,squared_error,absolute_error
0,5,600,725,-125,15625,125
1,7,900,905,-5,25,5
2,3,550,545,5,25,5
3,3,500,545,-45,2025,45
4,2,400,455,-55,3025,55
5,7,950,905,45,2025,45
6,3,540,545,-5,25,5
7,10,1200,1175,25,625,25
8,6,900,815,85,7225,85
9,4,550,635,-85,7225,85


In [5]:
# Gözlem sayısını al
n = len(df)

# MSE: hata karelerinin ortalaması
mse = df["squared_error"].sum() / n

# RMSE: MSE'nin karekökü (hatayı maaş biriminde ifade eder)
rmse = np.sqrt(mse)

# MAE: mutlak hataların ortalaması
mae = df["absolute_error"].sum() / n

# Hesaplanan metrikleri yazdır
print(f"MSE  : {mse:.4f}")
print(f"RMSE : {rmse:.4f}")
print(f"MAE  : {mae:.4f}")

MSE  : 4438.3333
RMSE : 66.6208
MAE  : 54.3333


In [6]:
# sklearn ile MSE hesaplayarak manuel sonucu doğrula
mse_sklearn = mean_squared_error(df["salary"], df["salary_pred"])

# sklearn MSE üzerinden RMSE hesapla
rmse_sklearn = np.sqrt(mse_sklearn)

# sklearn ile MAE hesapla
mae_sklearn = mean_absolute_error(df["salary"], df["salary_pred"])

# sklearn sonuçlarını yazdır
print(f"sklearn MSE  : {mse_sklearn:.4f}")
print(f"sklearn RMSE : {rmse_sklearn:.4f}")
print(f"sklearn MAE  : {mae_sklearn:.4f}")

sklearn MSE  : 4438.3333
sklearn RMSE : 66.6208
sklearn MAE  : 54.3333


## Sonuç Özeti

| Metrik | Değer |
|--------|-------|
| Model denklemi | $\hat{y} = 275 + 90x$ |
| MSE | 4438.3333 |
| RMSE | 66.6208 |
| MAE | 54.3333 |

**Yorum:**
- Model ortalama olarak yaklaşık **54.33** birim mutlak hata (MAE) yapmaktadır.
- RMSE ≈ **66.62** olduğundan, büyük hataların da etkisiyle ortalama sapma biraz daha yüksektir.
- MSE büyük hataları karesel olarak cezalandırdığı için sayısal değeri daha büyüktür; yorumlamada genellikle RMSE ve MAE tercih edilir.